File: DP/dijkstra_algo.java
class Solution {
    
    public int[] dijkstra(int V, int[][] edges, int src) {
        // code here
        Graph gh = new Graph(V);
        for(int i = 0; i<edges.length; i++){
            int node1 = edges[i][0];
            int node2 = edges[i][1];
            int weight = edges[i][2];
            gh.addEdge(node1, node2, weight);
        }
        int[] result = gh.dijkstra(src);
        return result;
    }
}

class Graph{
    static class Pair{
        int node;
        int weight;
        
        Pair(int node, int weight){
            this.node = node;
            this.weight = weight;
        }
        
    }
    private Map<Integer, List<Pair>> adjList;
    private int numOfNodes;
    
    public Graph(int n){
        numOfNodes = n;
        adjList = new HashMap<>();
    }
    
    public void addEdge(int node1, int node2, int weight){
        adjList.computeIfAbsent(node1, k-> new ArrayList<>()).add(new Pair(node2, weight));
        adjList.computeIfAbsent(node2, k-> new ArrayList<>()).add(new Pair(node1, weight));
    }
    public int[] dijkstra(int src){
        PriorityQueue<int []> pq = new PriorityQueue<>((a,b)-> Integer.compare(a[0], b[0]));
        int[] dist = new int[numOfNodes];
        Arrays.fill(dist, Integer.MAX_VALUE);
        int source = src;
        dist[source] = 0;
        pq.offer(new int[]{0, source});
        int[] result = new int[numOfNodes];
        while(!pq.isEmpty()){
            int[] curr = pq.poll();
            int currDist = curr[0];
            int node = curr[1];
            //skip outdated entries
            if(currDist > dist[node]){
                continue;
            }
            List<Pair> neighbors = adjList.getOrDefault(node, new ArrayList<>());
            for(Pair nbr : neighbors){
                int v = nbr.node;
                int weight = nbr.weight;
                if( dist[v] > dist[node] + weight){
                    dist[v] = dist[node] + weight;
                    pq.offer(new int[]{dist[v], v});
                }
            }
        }
       
        return dist;
    }
}


public class Main {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        int V = sc.nextInt();
        int E = sc.nextInt();
        int[][] edges = new int[E][3];
        for(int i = 0 i<E; i++){
            int u = sc.nextInt();
            int v = sc.nextInt();
            int wt = sc.nextInt();

            edges[i] = new int[]{u, v, wt};
        }
        Solution s = new Solution(V, edges, 2);

        sc.close();
    }
}


// V = 3, edges[][] = [[0, 1, 1], [1, 2, 3], [0, 2, 6]], src = 2

File: DP/dijkstra_algo.js
class Heap {
    constructor(compare) {
        this.heap = [];
        this.compare = compare;
    }

    size() {
        return this.heap.length;
    }

    peek() {
        return this.heap[0];
    }

    push(val) {
        this.heap.push(val);
        this.bubbleUp(this.heap.length - 1);
    }

    pop() {
        if (this.heap.length === 1) return this.heap.pop();

        const top = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown(0);

        return top;
    }

    bubbleUp(i) {
        while (i > 0) {
            let p = Math.floor((i - 1) / 2);

            if (this.compare(this.heap[p], this.heap[i])) break;

            [this.heap[p], this.heap[i]] =
                [this.heap[i], this.heap[p]];

            i = p;
        }
    }

    bubbleDown(i) {
        const n = this.heap.length;

        while (true) {
            let best = i;
            let left = 2 * i + 1;
            let right = 2 * i + 2;

            if (
                left < n &&
                !this.compare(this.heap[best], this.heap[left])
            )
                best = left;

            if (
                right < n &&
                !this.compare(this.heap[best], this.heap[right])
            )
                best = right;

            if (best === i) break;

            [this.heap[i], this.heap[best]] =
                [this.heap[best], this.heap[i]];

            i = best;
        }
    }
}

function dijkstra(graph, source) {

    const n = graph.length;

    const dist = new Array(n).fill(Infinity);

    dist[source] = 0;

    const minHeap = new Heap(
        (a, b) => a.distance <= b.distance
    );

    minHeap.push({
        node: source,
        distance: 0
    });

    while (minHeap.size()) {

        const { node, distance } = minHeap.pop();

        // Ignore stale entries
        if (distance > dist[node]) continue;

        for (const [next, weight] of graph[node]) {

            const newDistance = distance + weight;

            if (newDistance < dist[next]) {

                dist[next] = newDistance;

                minHeap.push({
                    node: next,
                    distance: newDistance
                });
            }
        }
    }

    return dist;
}